src/indexing_slicing.rs

```diff
@@ -0,0 +1,54 @@
+use rustc::lint::*;
+use rustc::middle::const_eval::EvalHint::ExprTypeChecked;
+use rustc::middle::const_eval::{eval_const_expr_partial, ConstVal};
+use rustc::middle::ty::TyArray;
+use rustc_front::hir::*;
+use utils::span_lint;
+
+/// **What it does:** Check for usage of indexing or slicing.
+///
+/// **Why is this bad?** This may panic at runtime.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+///
+/// ```
+/// let x = [1,2,3,4];
+/// x[2];
+/// &x[0..2];
+/// ```
+declare_lint! {
+    pub INDEXING_SLICING,
+    Allow,
+    "indexing/slicing usage"
+}
+
+#[derive(Copy,Clone)]
+pub struct IndexingSlicing;
+
+impl LintPass for IndexingSlicing {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INDEXING_SLICING)
+    }
+}
+
+impl LateLintPass for IndexingSlicing {
+    fn check_expr(&mut self, cx: &LateContext, e: &Expr) {
+        if let ExprIndex(ref array, ref index) = e.node {
+            // Don't warn if we statically know that the index is within bounds
+            let ty = cx.tcx.expr_ty(array);
+            if let TyArray(_, size) = ty.sty {
+                // Constant index
+                let index = eval_const_expr_partial(cx.tcx, &index, ExprTypeChecked, None);
+                if let Ok(ConstVal::Uint(index)) = index {
+                    if size as u64 > index {
```

Comment:
You can implement the two lints together here in the same `check_expr`. One `LateLintPass` implementation is not limited to one lint.